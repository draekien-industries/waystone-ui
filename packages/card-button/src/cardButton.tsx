'use client';

import { forwardRef, type ForwardedRef, type ReactNode } from 'react';
import type {
  DimensionsAttributes,
  DisabledAttributes,
  LoadingAttributes,
  TestIdAttributes,
} from '@waystone/types';
import { Spinner } from '@waystone/spinner';
import { Overlay } from '@waystone/overlay';
import { cssSelectors, outlineCss } from '@waystone/css-presets';
import { Flex, type ThemeUICSSObject } from 'theme-ui';

export type CardButtonProps = {
  onClick?: (e: React.MouseEvent<HTMLButtonElement>) => void;
  children?: ReactNode;
} & LoadingAttributes &
  DisabledAttributes &
  DimensionsAttributes &
  TestIdAttributes;

const renderCardButton = (
  {
    children,
    onClick,
    loading,
    disabled,
    'data-testid': testId,
    ...rest
  }: CardButtonProps,
  ref?: ForwardedRef<HTMLButtonElement>
) => {
  const styles: ThemeUICSSObject = {
    position: 'relative',
    padding: 'medium',
    border: 'none',
    borderRadius: 'medium',
    color: 'text',
    backgroundColor: 'card',
    boxShadow: 'small',
    cursor: 'pointer',
    overflow: 'hidden',
    transition: 'all 200ms ease',
    ...outlineCss.base,
    ...rest,
    [cssSelectors.hover]: {
      backgroundColor: 'active',
      boxShadow: 'large',
    },
    [cssSelectors.active]: {
      boxShadow: 'medium',
      backgroundColor: 'active',
      ...outlineCss.active,
    },
    [cssSelectors.focusVisible]: {
      boxShadow: 'medium',
      backgroundColor: 'active',
      ...outlineCss.focused,
    },
    [cssSelectors.disabled]: {
      color: 'text-disabled',
      backgroundColor: 'muted',
      cursor: 'not-allowed',
    },
  };

  return (
    <button
      ref={ref}
      sx={styles}
      onClick={onClick}
      type="button"
      disabled={disabled || loading}
      data-testid={testId}
    >
      {loading && (
        <Overlay.Absolute>
          <Flex
            sx={{
              justifyContent: 'center',
              alignItems: 'center',
              height: '100%',
            }}
          >
            <Spinner />
          </Flex>
        </Overlay.Absolute>
      )}
      {children}
    </button>
  );
};

export const CardButton = forwardRef(renderCardButton);
