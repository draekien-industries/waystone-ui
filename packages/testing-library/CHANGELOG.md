# @waystone/testing-library

## 9.2.0

### Minor Changes

- 21db598: chore: bump deps and add documentation

### Patch Changes

- Updated dependencies [21db598]
  - @waystone/theme@9.3.0

## 9.1.6

### Patch Changes

- 41664e2: fix: use type imports to improve bundling
- Updated dependencies [41664e2]
  - @waystone/theme@9.2.5

## 9.1.5

### Patch Changes

- 856ba2b: fix: missing dependencies when installing package
- Updated dependencies [856ba2b]
  - @waystone/theme@9.2.4

## 9.1.4

### Patch Changes

- 5b726aa: fix build and dependencies
- Updated dependencies [5b726aa]
  - @waystone/theme@9.2.3

## 9.1.3

### Patch Changes

- 7086019: Update dependencies
- Updated dependencies [7086019]
  - @waystone/theme@9.2.2

## 9.1.2

### Patch Changes

- @waystone/theme@9.2.1

## 9.1.1

### Patch Changes

- Updated dependencies [8f75dd6]
  - @waystone/theme@9.2.0

## 9.1.0

### Minor Changes

- 5c43357: Switch to pnpm and implement minimal syntax highlighter

### Patch Changes

- Updated dependencies [5c43357]
  - @waystone/theme@9.1.0

## 9.0.0

### Major Changes

- 4a19c42: Update package structure

  Created the `@waystone/ui` package for easier installation and created the `@waystone/testing-library`
  package for a preconfigured testing-library render function.

  Uninstall your `@waystone/*` packages, and run the command `yarn add @waystone/ui`
  to update to the latest version. You will need to update the imports to point
  to the new source.

### Patch Changes

- Updated dependencies [4a19c42]
  - @waystone/theme@9.0.0
