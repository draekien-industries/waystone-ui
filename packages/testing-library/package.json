{
  "name": "@waystone/testing-library",
  "version": "10.0.2",
  "private": true,
  "main": "./src/index.ts",
  "repository": {
    "type": "git",
    "url": "https://github.com/draekien-industries/waystone-ui"
  },
  "description": "The react component library used by draekien-industries",
  "author": "draekien-industries",
  "sideEffects": false,
  "license": "MIT",
  "files": [
    "src/**"
  ],
  "scripts": {
    "lint": "eslint --ext .tsx,.ts,.jsx,.js ./src/ --fix",
    "lint:release": "eslint --ext .tsx,.ts,.jsx,.js ./src/"
  },
  "devDependencies": {
    "@waystone/tsconfig": "workspace:*",
    "eslint-config-waystone": "workspace:*",
    "eslint-plugin-waystone": "workspace:*",
    "react": "^18.3.1",
    "react-dom": "^18.3.1"
  },
  "dependencies": {
    "@testing-library/jest-dom": "^6.6.3",
    "@testing-library/react": "^16.2.0",
    "@testing-library/user-event": "^14.6.0",
    "@theme-ui/color": "^0.17.1",
    "@theme-ui/css": "^0.17.1",
    "@waystone/theme-provider": "workspace:*",
    "jest": "^29.7.0",
    "jest-environment-jsdom": "^29.7.0",
    "theme-ui": "^0.17.1"
  },
  "publishConfig": {
    "access": "public"
  }
}
