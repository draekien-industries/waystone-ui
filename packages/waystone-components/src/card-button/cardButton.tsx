/** @jsxImportSource theme-ui */
import { alpha } from '@theme-ui/color';
import { getOutlineCss } from '@waystone/core';
import React from 'react';
import { ThemeUIStyleObject } from 'theme-ui';
import {
  CanBeFullWidth,
  CanDisable,
  CanLoad,
  HasHeight,
  HasWidth,
} from '../common';
import { useIsDarkMode } from '../hooks/useIsDarkMode';
import { Spinner } from '../spinner';

/**
 * The props that can be passed to the CardButton component.
 */
export interface CardButtonProps
  extends HasWidth,
    HasHeight,
    CanDisable,
    CanLoad,
    CanBeFullWidth {
  /** A custom onClick handler that will be executed when the button is clicked. */
  onClick?: (e: React.MouseEvent<HTMLButtonElement>) => void;
  /** Should the card button render without any dropshadow. */
  noShadow?: boolean;
  /** The content to render inside the card button. */
  children: React.ReactNode;
}

/**
 * A button that shares the same styling as a card.
 * @param props - The props that can be passed to the CardButton component.
 * @param props.children - The content to render inside the card button.
 * @param props.onClick - A custom onClick handler that will be executed when the button is clicked.
 * @param props.noShadow - Should the card button render without any dropshadow.
 * @param props.width - The width of the card button.
 * @param props.minWidth - The minimum width of the card button.
 * @param props.maxWidth - The maximum width of the card button.
 * @param props.height - The height of the card button.
 * @param props.minHeight - The minimum height of the card button.
 * @param props.maxHeight - The maximum height of the card button.
 * @param props.disabled - Should the card button be disabled.
 * @param props.loading - Should the card button be loading.
 * @param props.fullWidth - Should the card button be full width.
 * @returns The rendered card button.
 */
export const CardButton = ({
  children,
  fullWidth,
  width,
  minWidth,
  maxWidth,
  noShadow,
  disabled,
  loading,
  onClick,
  ...rest
}: CardButtonProps) => {
  const darkMode = useIsDarkMode();
  const outline = getOutlineCss();

  const cardCss: ThemeUIStyleObject = {
    position: 'relative',
    paddingX: 'md',
    paddingY: 'md',
    border: 'none',
    borderRadius: 'md',
    color: 'text',
    width: fullWidth ? '100%' : width || 'auto',
    minWidth: !fullWidth && minWidth,
    maxWidth: !fullWidth && maxWidth,
    backgroundColor: darkMode ? 'b-600' : 'b-50',
    boxShadow: !noShadow && 'md',
    cursor: 'pointer',
    overflow: 'hidden',
    textAlign: 'left',
    transition: 'all 200ms',
    ...outline.base,
    ...outline.focused,
    ...rest,
    ':enabled:hover': {
      backgroundColor: darkMode ? 'b-500' : 'b-100',
    },
    ':enabled:active': {
      backgroundColor: darkMode ? 'b-400' : 'b-200',
      ...outline.active[':enabled:active'],
    },
    ':disabled': {
      color: 'b-400',
      backgroundColor: 'muted',
      cursor: 'not-allowed',
    },
  };

  return (
    <button
      sx={cardCss}
      onClick={onClick}
      type="button"
      disabled={disabled || loading}
    >
      {loading && <Spinner overlay />}
      {children}
    </button>
  );
};
