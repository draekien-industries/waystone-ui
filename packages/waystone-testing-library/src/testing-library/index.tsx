'use client';

import type { RenderOptions, RenderResult } from '@testing-library/react';
import { render } from '@testing-library/react';
import { Waystone } from '@waystone/theme-provider';
import type { PropsWithChildren } from 'react';
import type React from 'react';
import { useEffect } from 'react';
import { useColorMode } from 'theme-ui';

const TestProvider = ({ children }: PropsWithChildren) => (
  <Waystone>{children}</Waystone>
);

const DarkModeWrapper = ({ children }: PropsWithChildren) => {
  const [, setColorMode] = useColorMode();

  useEffect(() => {
    setColorMode('dark');
  }, [setColorMode]);

  return <div>{children}</div>;
};

const themedRender: (
  ui: React.ReactElement,
  options?: RenderOptions
) => RenderResult = (ui, options?) =>
  render(ui, { wrapper: options?.wrapper || TestProvider, ...options });

export * from '@testing-library/react';
export * from '@testing-library/user-event';

export { DarkModeWrapper, themedRender as render };
