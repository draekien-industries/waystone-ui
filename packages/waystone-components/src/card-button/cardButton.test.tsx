'use client';

import { fireEvent, render, screen } from '@waystone/testing-library';
import { CardButton } from './cardButton';

describe('TThe CardButton component', () => {
  it('Matches snapshot', () => {
    const { container } = render(<CardButton>card</CardButton>);

    expect(container).toMatchSnapshot();
  });

  it('Invokes the onClick callback when clicked', async () => {
    const handleClick = jest.fn();

    render(<CardButton onClick={handleClick}>card</CardButton>);

    const button = await screen.findByText('card');

    fireEvent.click(button);

    expect(handleClick).toHaveBeenCalledTimes(1);
  });
});
