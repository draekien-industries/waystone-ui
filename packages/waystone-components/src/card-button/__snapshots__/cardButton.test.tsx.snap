// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TThe CardButton component Matches snapshot 1`] = `
<div>
  <button
    class="css-5143ny-CardButton"
    type="button"
  >
    card
  </button>
</div>
`;
